src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ExpressionTextCache.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures
+{
+    /// <summary>
+    /// This class holds the cache for the expression text that is computed by ExpressionHelper.
+    /// </summary>
+    public class ExpressionTextCache
+    {
+        public ConcurrentDictionary<LambdaExpression, string> Entries { get; } = 
+            new ConcurrentDictionary<LambdaExpression, string>(LambdaExpressionComparer.Instance);
+
+        private class LambdaExpressionComparer : IEqualityComparer<LambdaExpression>
+        {
+            public static readonly LambdaExpressionComparer Instance = new LambdaExpressionComparer();
+
+            public bool Equals(LambdaExpression lambdaExpression1, LambdaExpression lambdaExpression2)
+            {
+                if (ReferenceEquals(lambdaExpression1,lambdaExpression2))
+                {
+                    return true;
+                }
+                // We will cache only pure member access expressions. Hence we compare two expressions 
+                // to be equal only if they are identical member access expressions.
+                var expression1 = lambdaExpression1.Body;
+                var expression2 = lambdaExpression2.Body;
+                var isEquivalent = false;
+
+                while (true)
+                {
+                    if (expression1 == null || expression2 == null)
```

Comment:
Just wondering: Are either `((LambdaExpression)x).Body == null` or `((MemberExpression)y).Expression == null` actual cases? (Realize `ExpresisonHelper` uses `part != null` for its loop condition but am not sure that guard is necessary.)

If this can really happen, `expression1 == null && expression2 == null` should be considered equivalent.